Reader: Refactor filter sheet

dvdchr opened the pull request:
Closes #22211

> [!NOTE]
> This merges into `feature/reader-navigation-filter`.

This PR covers some groundwork changes for the upcoming filter pill buttons and updates the look of the Filter sheet. My original work in progress has become too large and I decided to split the changes, as I'm also in the middle of rebasing my ongoing work on top of your changes at #22260.

Included in this PR: 

- Refactor `FilterSheetViewController`; merge `FilterSheetView` into the view controller and remove any tab management code.
- Update the cell style for 'filter by blog' and 'filter by tag'. ⚠️ **Note that this excludes the Edit button.** I'll add it through a separate PR.
- Move the `FilterProvider` management to the tab view model.
- Show a simple Text element to indicate how many filters are available for the current stream.
- Rename `shouldHideButtons` to `shouldHideStreamFilters` to fit better with the new navigation hierarchy.
---

Apart from this change, in summary, here's how the filter implementation is going to change (which would be covered in another 2-3 follow-up PRs):

- The filter sheet will only hold one instance of `FilterProvider`. 
- The `FilterProvider`s will serve as the model for the filter buttons—Imagine pulling out the tabs within the current Filter sheet and turning them into the new filter buttons. 
- Previously, the filter sheet fetches the list of blogs (or tags) _after_ the sheet is loaded. In the new design, this needs to happen every time the user switches the stream. 

## To test

- Launch the Jetpack app and go to Reader.
- The filter sheet is now temporarily placed in the Settings button. Tap the Settings button while in the Following tab to open the filter sheet for tags.
- 🔎  Verify that the filter sheet appears and shows 'Filter by tag'.
- 🔎  Verify that selecting the tag filters the current stream.
- Switch to an organization tab and tap the Settings button.
- 🔎  Verify that the filter sheet appears and shows 'Filter by blog'.
- 🔎  Verify that selecting the blog filters the current stream.

> [!WARNING]
> There's currently no way to reset the filter for now as the entry point is just a makeshift to test the sheet implementation. I'll remove the test code when this is approved.

## Regression Notes
1. Potential unintended areas of impact
Should be none. The feature is unreleased and there will be further tests on the main feature branch `feature/reader-navigation-filter`.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manually tested the changes.

3. What automated tests I added (or what prevented me from doing so)
N/A.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [x] Portrait and landscape orientations.
- [x] Light and dark modes.
- [x] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [x] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

dvdchr committed:
Refactor FilterSheetViewController

In the new design, the sheet will only handle one filter "tab"
at any given time. This refactor work removes the code related to
tab management within the filter sheet, and simplifies the code by
merging the view implementation into the view controller.

dvdchr committed:
Update table view cell style in the filter sheet

wargcm commented on WordPress/Classes/ViewRelated/Reader/Filter/FilterSheetViewController.swift:
```diff
@@ -1,14 +1,86 @@
+    }()
+
+    private lazy var emptyView: EmptyActionView = {
+        let view = EmptyActionView(tappedButton: tappedEmptyAddButton)
```
This is causing a retain cycle:
![Screenshot 2024-01-02 at 3 12 19 PM](https://github.com/wordpress-mobile/WordPress-iOS/assets/2454408/4275bb03-92b6-4843-a419-f113caff490b)

```suggestion
        let view = EmptyActionView { [weak self] in
            self?.tappedEmptyAddButton()
        }
```

This may be an existing issue already, since you're copying a lot of the code from the previous file.

dvdchr replied:
Whoops, I applied your suggestion but then force pushed from my local changes, causing this to be overwritten... I've re-applied this again here: https://github.com/wordpress-mobile/WordPress-iOS/pull/22273/commits/05504dd2f82d620d8f6b6b774d9d0eeeba7a97f3

wargcm reviewed (commented):
Could you share your test code to present the view? I was having issues with the view on iPad, and I think it might be because of my test code.

It also appears that there's a couple failing tests from our changes. Looks like in both cases, it's due to accessing the `tabItems` array. Do you want me to take a look at `testStatsWidgetsDeletedAfterDefaultWPAccountRemoved` since it seems related to the menu I added?

dvdchr committed:
Move FilterProvider creation to the view model

In addition, this also:

* Reloads the filter list immediately once the FilterProvider is loaded.
* Shows a simple Text element to indicate how many filters are available
  for the current stream.
* Renames `shouldHideButtons` to `shouldHideStreamFilters` to fit
  better with the new navigation hierarchy.

dvdchr committed:
Temporarily adjust filter sheet creation logic

dvdchr committed:
Patch failing test, prevent index out of bounds

dvdchr committed:
Fix retain cycle on EmptyActionView

dvdchr commented:
> It also appears that there's a couple failing tests from our changes. Looks like in both cases, it's due to accessing the tabItems array. Do you want me to take a look at `testStatsWidgetsDeletedAfterDefaultWPAccountRemoved` since it seems related to the menu I added?

Hmm, I tried to fix this in https://github.com/wordpress-mobile/WordPress-iOS/pull/22273/commits/a6315e87381e29777377619e57d42204e37ee361. It crashes locally for me when running the full suite because the navigation menu tries to access the first index on an empty array. For now, I've made it so ReaderNavigationMenu works with an optional `selectedItem` and I just hide the button if there isn't one.

However, even after fixing the above it still fails. There is a small delay when writing to the cache for the widgets data. I've added a 50ms delay before asserting that the data exists in the cache, but let me know if you have better ideas!

> Could you share your test code to present the view? 

Ah, it looks like the test code got overwritten when I rebased the `feature/reader-navigation-filter` on the latest `trunk` last week, and then rebased this PR again on `feature/reader-navigation-filter`. 

Instead of re-adding the test code, in https://github.com/wordpress-mobile/WordPress-iOS/pull/22273/commits/63047135bf35f2f2c00050c885c35015dedee553 I've cherry-picked my commit from an upcoming branch so that it shows a `Text` beside the `ReaderNavigationButton` depending on how many stream filters are available.

<img width="389" alt="Screenshot 2024-01-03 at 19 06 24" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1299411/1c5d4081-6ce0-48f1-aa58-8f5792f1ae66">

They are just simple texts right now, but the next PR will show the filter sheet. It's a bit messy because I'll have to fix the presentation unit tests.

wargcm reviewed (approved):
Thanks for addressing those issues 👍. Since this one is a bit difficult to test, I'm going to approve it based off my previous test code. I'll do a more in-depth test on the UI in one of your futures PRs.
